[Async CC] Add constant "pointer" for async func.

nate-chandler opened the pull request:
An AsyncFunctionPointer, defined in Task.h, is a struct consisting of two i32s: (1) the relative address of the async function and (2) the size of the async context to be allocated when calling that function.

Here, such structs are emitted for every async SILFunction that is emitted.  Furthermore, pointers to these new structs rather than pointers directly to async functions are passed around.  The async context size is now read out of these function pointers and used to allocate the async context.  This will enable the size constants to be rewritten by an LLVM pass with the ideal storage space to minimize additional allocations.

The following tasks need to be completed before merge:
- [x] Avoiding AsyncFunctionPointers for partial apply forwarders.
- [x] Using SILFunction as the value for the AsyncFunctionPointer LinkEntity.
- [x] Using a LinkEntity for TBDGen of AsyncFunctionPointers.
- [x] Avoiding extracting function pointers from AsyncFunctionPointers during thin_to_thick and convert_function lowering.
- [x] Pointer authentication for function pointers.

The following related task will be completed in follow-on PRs:
- Referring directly to async functions when particular callees are known rather than loading the function pointers from AsyncFunctionPointer structs.

nate-chandler committed:
[NFC] Corrected typo.

ClassMetadataLayout::getOffsett -> ClassMetadataLayout::getOffset

nate-chandler committed:
[Async CC] Add AsyncFunctionPointer to vtable.

nate-chandler committed:
[Async CC] Add AsyncFunctionPointer to wtable.

DougGregor reviewed (commented):
Very cool!

nate-chandler commented on lib/IRGen/GenCall.cpp:
```diff
@@ -91,9 +91,10 @@ AsyncContextLayout irgen::getAsyncContextLayout(IRGenModule &IGM,
+AsyncContextLayout
+irgen::getAsyncContextLayout(IRGenModule &IGM, CanSILFunctionType originalType,
+                             CanSILFunctionType substitutedType,
+                             SubstitutionMap substitutionMap) {
```
Sorry, unrelated.  I'll break it out of this PR.

nate-chandler commented on lib/IRGen/GenCall.cpp:
```diff
@@ -4266,7 +4274,12 @@ Callee irgen::getSwiftFunctionPointerCallee(
-  FunctionPointer fn(fnPtr, authInfo, sig);
+  FunctionPointer fn((calleeInfo.OrigFnType->isAsync() &&
+                      calleeInfo.OrigFnType->getRepresentation() !=
+                          SILFunctionTypeRepresentation::Thick)
```
Just a temporary hack, I've been working on removing it.

nate-chandler commented on lib/IRGen/GenCall.cpp:
```diff
@@ -4282,32 +4295,56 @@ Callee irgen::getCFunctionPointerCallee(IRGenFunction &IGF,
+  if (kind == KindTy::AsyncFunctionPointer) {
+    assert(fnType->getRepresentation() != SILFunctionTypeRepresentation::Thick);
+  }
+  return forDirect(kind, fnPtr, IGM.getSignature(fnType));
```
Something like this?

```
// Have to dig the size out of the AsyncFunctionPointer so that it can be rewritten downstream
%afp = `AsyncFunctionPointer(callee)`
%size_addr = getelementptr %afp, 0, 1
%size = load %size_addr
%context = call @swift_task_alloc(%task, %size)
// But we don't have to dig the function pointer out of the AsyncFunctionPointer because we already know what it is
call @callee(%task, %executor, %context)
```

nate-chandler committed:
[TBDGen] Add async function pointers.

In order to call async functions, instances of the AsyncFunctionPointer
struct must be used.  If those functions are exported from a module, the
AsyncFunctionPointer by means of which the function is to be called must
be exported as well.

For now, the symbol is exported by manually appending the relevant
suffix to the mangled name of the function.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please clean test windows platform

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please clean test windows

aschwaighofer commented on lib/IRGen/GenFunc.cpp:
```diff
@@ -2173,8 +2180,19 @@ Optional<StackAddress> irgen::emitFunctionPartialApplication(
+    // TODO: Rather than creating an AsyncFunctionPointer in this case, use a
+    //       bare function pointer and teach apply/partial_apply lowering to
+    //       handle thick async functions differently. 
+    //       [lowering_thick_async_functions_without_sizes]
```
I think this will cause TBDGen verification errors because we take the visibility of the async function pointer global off the `SILFn` but the partial apply forwarding thunk is internal and the async function pointer for it should also be internal.

```
public func callAsyncClosure(_ closure : () async->()) async {                  
  await closure()                                                               
}                                                                               
                                                                                
public class SomeClass {}

public func useAsyncClosure(_ c: SomeClass) async {                             
  func closure() async {                                                        
    print(c)                                                                    
  }                                                                             
  await callAsyncClosure(closure)                                               
} 
```

I know this is ultimately going away but in the meantime I think you could work around this by passing a null SILFunction.

nate-chandler replied:
Just updated the PR, and AsyncFunctionPointers for partial apply forwarders are no longer being emitted.

nate-chandler committed:
[IRGen] Added LinkEntities for AsyncFunctionPointer.

Two LinkEntities are needed to enable the construction during both IRGen
and TBDGen.

nate-chandler committed:
[Async CC] Add constant "pointer" for async func.

An AsyncFunctionPointer, defined in Task.h, is a struct consisting of
two i32s: (1) the relative address of the async function and (2) the
size of the async context to be allocated when calling that function.

Here, such structs are emitted for every async SILFunction that is
emitted.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please clean test windows

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test

nate-chandler committed:
To Revert: [Async CC] Disable execution tests on arm64e.

While the pointer auth details are worked out, to unblock other work,
disable the tests on arm64e where they will fail until those details are
worked out.

nate-chandler commented:
@swift-ci please test

nate-chandler commented:
@swift-ci please test
